### PR TITLE
Date refactorting to 5.5

### PR DESCRIPTION
**Pull request by gianarb**

### CODE REVIEW

**Comment by Ocramius** [comment]
Tests please!

**Comment by gianarb** [comment]
thanks @Ocramius can you help me to write a good tests to coverage this feature?
 it's already coverage IMO

**Comment by Ocramius** [comment]
> it's already coverage IMO

The particular path is not followed. You'd simply `$this->getMock(DateTimeInterface::class)` and work with that

**Comment by gianarb** [comment]
Why not?
https://github.com/zendframework/zend-validator/blob/master/test/DateTest.php#L85

DateTime implement DateTimeinterface

**Comment by Ocramius** [comment]
@gianarb but you are not testing that any implementation of `DateTimeInterface` passes the test.

**Comment by samsonasik** [comment]
@gianarb thanks for this ;)

**Comment by gianarb** [comment]
Little problem
http://3v4l.org/8GvgO

https://github.com/php/php-src/blob/4a2e40bb861bc3cf5fb6863e57486ed60316e97c/ext/date/php_date.c#L1961

Thanks @fntlnz :P

**Comment by Ocramius** [comment]
@gianarb as it stands, you'd probably have to test with `DateTime` and `DateTimeImmutable` then. I raised the question at http://news.php.net/php.internals/87217

**Comment by gianarb** [comment]
in this moment we have already wrote tests with DateTime, DateTime Immutable and stdClass

**Comment by samsonasik** [comment]
@Ocramius tests updated by @gianarb ;)

**Comment by Maks3w** [comment]
This is pending of RFC with a proposal of Interface removal. This will take days or weeks.

File: src/Date.php [hunk 1]
```diff
@@ -127,8 +127,7 @@ public function isValid($value)
      */
     protected function convertToDateTime($param, $addErrors = true)
     {
-        // @TODO: when minimum dependency will be PHP 5.5, we can only keep check against DateTimeInterface
-        if ($param instanceof DateTime || $param instanceof DateTimeInterface) {
+        if ($param instanceof DateTimeInterface) {
```
Comment (Ocramius) [inline src/Date.php]: As per discussion with the internal folks (sigh), this should probably be:

```php
if ($param instanceof DateTime || $param instanceof DateTimeImmutable) {
```

According to internal folks, `DateTimeInterface` is not to be used.

**Comment by samsonasik** [comment]
@Ocramius , @gianarb has updated to instanceof DateTimeImmutable ;)

**Comment by samsonasik** [comment]
@Ocramius @gianarb I think the PR title can be updated

**Pull request merged**